### ISSUE TITLE
mask_and_reduce_metric does not support area-weighted averaging over grid points

### ISSUE THREAD

**Issue by prajwal-tech07** [mllam/neural-lam#212]
### **Problem**

All six metrics in metrics.py (mse, mae, wmse, wmae, nll, crps_gauss) reduce over the spatial grid dimension using a simple unweighted mean:

     if average_grid:  # Reduce grid first
          metric_entry_vals = torch.mean(metric_entry_vals, dim=-2)

This implicitly assumes that every grid point represents the same geographic area. That assumption is invalid for:

- **Latitude–longitude grids** — grid cells near the poles cover significantly less surface area than cells near the equator (e.g., a 1°×1° cell at 60°N has   
   roughly half the area of one at the equator).
- **LAM domains spanning wide latitude ranges** — even conformal projections do not perfectly equalize cell areas.
- **Global grids** — distortion becomes extreme, leading to strong polar over-representation.

As a result, reported RMSE/MAE/CRPS values are biased toward high-latitude regions, potentially misleading model evaluation and comparison.

### **Where the Problem Propagates**

Because all metrics call mask_and_reduce_metric, the unweighted averaging affects:

- Training loss via self.loss(...) in training_step
- Validation loss in validation_step
- Test metrics (MSE, MAE, etc.) in test_step

The issue originates from a single reduction point: mask_and_reduce_metric.
Therefore, it can be fixed centrally

### Proposed Solution

Add an optional weights parameter to mask_and_reduce_metric and propagate it through all metric functions.

**Updated Reduction Logic**

    def mask_and_reduce_metric(metric_entry_vals, mask, average_grid, weights=None):

          # Apply mask
          if mask is not None:
                 metric_entry_vals = metric_entry_vals[..., mask]
                 if weights is not None:
                        weights = weights[mask]

          # Reduce grid dimension
          if average_grid:
                 if weights is not None:
                        metric_entry_vals = (
                             (metric_entry_vals * weights.unsqueeze(...)).sum()
                               / weights.sum()
                             )
                else:
                        metric_entry_vals = torch.mean(metric_entry_vals, dim=-2)

### Metric Functions

Each metric (mse, mae, wmse, wmae, nll, crps_gauss) should:

- Add weights=None as an optional argument
- Pass weights through to mask_and_reduce_metric
- Keep the DEFINED_METRICS dictionary unchanged

###Model Integration

On the model side, ARModel can precompute area weights (e.g., cos(latitude)) from the datastore’s coordinate data and register them as a buffer. These weights can then be passed into loss and metric calls.

When weights=None (default), behavior remains fully backward compatible.

### Expected Outcome

- Correct area-weighted evaluation for global and wide-latitude grids
- No breaking changes
- Single-point fix affecting training, validation, and testing consistently
- Improved scientific correctness of reported metrics

 **I would be happy to work on this issue and implement the proposed changes.
 Please assign it to me if it looks good.**

**Comment by prajwal-tech07** [comment]
@sadamov can you assign it to me also?

**Comment by prajwal-tech07** [comment]
@sadamov , I have resolved this issue and pushed it,you can review it ones before merging ,and can close the issue

**Comment by joeloskarsson** [comment]
Just a note to myself and any other contributor looking at this: Work on this is actively ongoing in #258

**Comment by sadamov** [comment]
Closing as the area-weighted grid-point averaging work is now tracked in the consolidated #650 (global forecasting / spherical domain support). The implementation will land via the curated PR there.
